File: anime/src/en/hentaitorrents.js
// --- METADATA ---
const mangayomiSources = [{
    "name": "HentaiTorrents",
    "id": 5928374928374291,
    "lang": "en",
    "baseUrl": "https://www.hentaitorrents.com",
    "apiUrl": "",
    "iconUrl": "https://www.google.com/s2/favicons?sz=256&domain=https://www.hentaitorrents.com",
    "typeSource": "single",
    "itemType": 1,
    "version": "1.0.1",
    "pkgPath": "anime/src/en/hentaitorrents.js"
}];

// --- CLASS ---
class DefaultExtension extends MProvider {
    constructor() {
        super();
        this.client = new Client();
    }

    getBaseUrl() {
        return this.source.baseUrl;
    }

    getHeaders(url) {
        return {
            "Referer": this.getBaseUrl(),
            "Origin": this.getBaseUrl(),
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36"
        };
    }

    parseListing(doc) {
        const list = [];
        const items = doc.select("div.image-wrapper");
        for (const item of items) {
            const name = item.selectFirst("a.overlay").text;
            let link = item.selectFirst("a.overlay").getHref;
            link = new URL(link, this.getBaseUrl()).href;
            const imageUrl = item.selectFirst("img").getSrc;
            list.push({ name, imageUrl, link });
        }
        return list;
    }

    async getPopular(page) {
        const url = page === 1 ? this.getBaseUrl() : `${this.getBaseUrl()}/page/${page}`;
        const res = await this.client.get(url, this.getHeaders(url));
        const doc = new Document(res.body);
        const list = this.parseListing(doc);
        return { list, hasNextPage: list.length > 0 };
    }

    async getLatestUpdates(page) {
        return this.getPopular(page);
    }

    async search(query, page, filters) {
        const url = `${this.getBaseUrl()}/?q=${encodeURIComponent(query)}&f_page=${page}`;
        const res = await this.client.get(url, this.getHeaders(url));
        const doc = new Document(res.body);
        const list = this.parseListing(doc);
        return { list, hasNextPage: list.length > 0 };
    }

    async getDetail(url) {
        const res = await this.client.get(url, this.getHeaders(url));
        const doc = new Document(res.body);
        const name = doc.selectFirst("h1").text;
        const imageUrl = doc.selectFirst("div.container > img").getSrc;
        const link = url;
        let description = "No description available.";
        const genre = [];
        const articleContentElement = doc.selectFirst("div.article-content");
        if (articleContentElement) {
            const articleHtml = articleContentElement.html;
            const descMatch = articleHtml.match(/<b>\s*Description\s*<\/b>:(.*)/si);
            if (descMatch && descMatch[1]) {
                description = descMatch[1].replace(/<br\s*\/?>/gi, '\n').replace(/<[^>]*>/g, '').trim();
            } else {
                description = articleContentElement.text.trim();
            }
            const articleText = articleContentElement.text;
            const genreMatch = articleText.match(/Genre\s*:\s*([^\r\n]+)/);
            if (genreMatch && genreMatch[1]) {
                genre.push(...genreMatch[1].split(',').map(g => g.trim()));
            }
        }
        const status = 1;
        const chapters = [];
        const downloadElement = doc.selectFirst("div.download-container a.download-button");
        if (downloadElement) {
            let epUrl = downloadElement.getHref;
            epUrl = new URL(epUrl, this.getBaseUrl()).href;
            chapters.push({
                name: "Download Torrent",
                url: epUrl,
            });
        }
        return { name, imageUrl, description, link, status, genre, chapters };
    }

    async getVideoList(url) {
        const res = await this.client.get(url, this.getHeaders(url));
        const doc = new Document(res.body);
        const streams = [];
        const torrentLinkElement = doc.selectFirst("a.download-button");
        if (torrentLinkElement) {
            const torrentUrl = torrentLinkElement.getHref;
            streams.push({
                url: torrentUrl,
                originalUrl: torrentUrl,
                quality: "Torrent",
                headers: this.getHeaders(torrentUrl)
            });
        }
        return streams;
    }

    getFilterList() {
        return [];
    }

    getSourcePreferences() {
        return [];
    }
}

File: anime/src/en/hentaitorrents.dart
import 'package:mangayomi/bridge_lib.dart';

class HentaiTorrents extends MProvider {
  HentaiTorrents({required this.source});

  MSource source;

  final Client client = Client();

  @override
  Future<MPages> getPopular(int page) async {
    final url = page == 1 ? getBaseUrl() : "${getBaseUrl()}/page/$page";
    final res = (await client.get(Uri.parse(url))).body;
    return parseAnimeList(res);
  }

  @override
  Future<MPages> getLatestUpdates(int page) async {
    // The site's homepage (popular) also serves as the latest updates.
    return getPopular(page);
  }

  @override
  Future<MPages> search(String query, int page, FilterList filterList) async {
    final url =
        "${getBaseUrl()}/?q=${Uri.encodeComponent(query)}&f_page=$page";
    final res = (await client.get(Uri.parse(url))).body;
    return parseAnimeList(res);
  }

  @override
  Future<MManga> getDetail(String url) async {
    MManga anime = MManga();
    final res = (await client.get(Uri.parse(url))).body;
    final document = parseHtml(res);

    anime.name = document.selectFirst("div.container > h1")?.text ?? "";
    anime.imageUrl = document.selectFirst("div.container > img")?.getSrc ?? "";

    final articleContentElement = document.selectFirst("div.article-content");
    if (articleContentElement != null) {
      final articleText = articleContentElement.text;

      final descMatch = RegExp(r'Description\s*:\s*([\s\S]*)', caseSensitive: false).firstMatch(articleText);
      anime.description = descMatch?.group(1)?.trim() ?? articleText.trim();

      final genreMatch = RegExp(r'Genre\s*:\s*([^\r\n]+)').firstMatch(articleText);
      if (genreMatch != null) {
        final genreString = genreMatch.group(1) ?? "";
        anime.genre = genreString.split(',').map((e) => e.trim()).toList();
      }
    }

    List<MChapter> chapters = [];
    final downloadElement = document.selectFirst("div.download-container a.download-button");
    if (downloadElement != null) {
      final epUrl = downloadElement.getHref;
      if (epUrl.isNotEmpty) {
        chapters.add(
          MChapter(
            name: "Download Torrent",
            url: getAbsoluteUrl(epUrl),
          ),
        );
      }
    }
    anime.chapters = chapters;
    anime.status = MStatus.completed;

    return anime;
  }

  @override
  Future<List<MVideo>> getVideoList(String url) async {
    final res = (await client.get(Uri.parse(url))).body;
    final document = parseHtml(res);
    final torrentLinkElement = document.selectFirst("div.container a.download-button");
    
    if (torrentLinkElement != null) {
      final torrentUrl = torrentLinkElement.getHref;
      if (torrentUrl.isNotEmpty) {
        return [
          MVideo(url: torrentUrl, originalUrl: torrentUrl, quality: "Torrent")
        ];
      }
    }
    return [];
  }

  MPages parseAnimeList(String res) {
    List<MManga> animeList = [];
    final document = parseHtml(res);

    final items = document.select("div.image-wrapper");
    for (var item in items) {
      MManga anime = MManga();
      anime.name = item.selectFirst("a.overlay")?.text ?? "";
      final link = item.selectFirst("a.overlay")?.getHref ?? "";
      if (link.isNotEmpty) {
        anime.link = getAbsoluteUrl(link);
      }
      anime.imageUrl = item.selectFirst("img")?.getSrc ?? "";
      if (anime.name.isNotEmpty && anime.link.isNotEmpty) {
        animeList.add(anime);
      }
    }

    return MPages(animeList, animeList.isNotEmpty);
  }
  
  String getBaseUrl() {
    return source.baseUrl;
  }
  
  String getAbsoluteUrl(String url) {
    if (url.startsWith("http")) {
      return url;
    }
    return "${getBaseUrl()}${url.startsWith('/') ? url : '/$url'}";
  }

  @override
  List<dynamic> getFilterList() {
    return [];
  }

  @override
  List<dynamic> getSourcePreferences() {
    return [];
  }
}

HentaiTorrents main(MSource source) {
  return HentaiTorrents(source: source);
}
